shared/src/main/scala/mlscript/JSBackend.scala

```diff
@@ -406,55 +423,45 @@ class JSBackend(pgrm: Pgrm) {
           // defs.name = name;
           // ```
           .concat(otherStmts.flatMap {
-            case Def(isRecursive, name, L(body)) =>
-              val translatedBody = translateTerm(body)
+            case Def(isRecursive, originalName, L(body)) =>
+              val translatedBody = translateTerm(body)(topLevelScope)
+              // `declName` menas the name used in the declaration.
+              // We allow define functions with the same name as classes.
               // Get a name that not conflicts with class names.
-              val tempName = (name #:: LazyList
-                .from(0)
-                .map { name + _.toString }).dropWhile { classNames contains _ }.head
-              if (tempName =/= name) {
-                letLhsAliasMap += name -> tempName
-              }
-              //
-              val shadowedName = resolveShadowName(tempName)
-              if (shadowedName === tempName) {
-                // Declare the name, assign and record the value.
-                // ```
-                // let <tempName> = <expr>;
-                // defs.<name> = <tempName>;
-                // ```
-                JSLetDecl(tempName, translatedBody) ::
-                  JSExprStmt(
-                    JSAssignExpr(
-                      JSMember(JSIdent(defResultObjName), name),
-                      JSIdent(tempName)
-                    )
-                  ) :: Nil
+              val declName = if (classNames contains originalName) {
+                val alias = topLevelScope allocate originalName
+                ctorAliasMap += originalName -> alias
+                alias
               } else {
-                // Re-assign and record the value as a new name:
-                // ```
-                // <tempName> = <expr>;
-                // defs["<name>@<number of shadow>"] = <tempName>;
-                // ```
-                JSExprStmt(JSAssignExpr(JSIdent(tempName), translatedBody)) ::
-                  JSExprStmt(
-                    JSAssignExpr(
-                      JSMember(JSIdent(defResultObjName), shadowedName),
-                      JSIdent(tempName)
-                    )
-                  ) :: Nil
+                originalName
               }
+              // We need to save a snapshot after each computation.
+              // The snapshot name will be same as the original name, or in the
+              // format of `originalName@n` if the original name is used.
+              val snapshotName = resolveShadowName(originalName)
+              (if (snapshotName === originalName) {
+                 // First time: `let <declName> = <expr>;`
+                 JSLetDecl(declName, translatedBody)
+               } else {
+                 // Not first time: `<declName> = <expr>;`
+                 JSExprStmt(JSAssignExpr(JSIdent(declName), translatedBody))
```

Comment:
Ok now I see that's what's causing https://github.com/hkust-taco/mlscript/issues/49#issuecomment-994556548. Incorrect logic here – a renaming should be used instead. I guess you'll fix it later and that's okay.

Reply:
Sure.